geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/SimpleFeatureSerializer.scala

```diff
@@ -0,0 +1,346 @@
+ * see @KryoFeatureSerializer
+ *
+ * @param sft
+ */
+class SimpleFeatureSerializer(sft: SimpleFeatureType) extends Serializer[SimpleFeature] {
+
+  import org.locationtech.geomesa.feature.kryo.SimpleFeatureSerializer._
+
+  val encodings = sftEncodings(sft)
+  val decodings = sftDecodings(sft)
+
+  override def write(kryo: Kryo, output: Output, sf: SimpleFeature): Unit = {
+    output.writeInt(VERSION, true)
+    output.writeString(sf.getID)
+    encodings.foreach(encode => encode(output, sf))
+  }
+
+  override def read(kryo: Kryo, input: Input, typ: Class[SimpleFeature]): SimpleFeature = {
+    input.readInt(true) // discard version info, currently only one version
+    val id = input.readString()
+    val values = Array.ofDim[AnyRef](sft.getAttributeCount)
+
+    decodings.foreach { case (decode, i) => values(i) = decode(input) }
+
+    new KryoSimpleFeature(id, sft, values)
+  }
+}
+
+/**
+ * Reads just the id from a serialized simple feature
+ */
+class FeatureIdSerializer extends Serializer[KryoFeatureId] {
+
+  override def write(kryo: Kryo, output: Output, id: KryoFeatureId): Unit = ???
+
+  override def read(kryo: Kryo, input: Input, typ: Class[KryoFeatureId]): KryoFeatureId = {
+    input.readInt(true) // discard version info, currently only one version
+    KryoFeatureId(input.readString())
+  }
+}
+
+/**
+ * Kryo serialization implementation for simple features - provides transformation during read
+ *
+ * @param sft
+ * @param decodeAs
+ */
+class TransformingSimpleFeatureSerializer(sft: SimpleFeatureType, decodeAs: SimpleFeatureType)
+    extends SimpleFeatureSerializer(sft) {
+
+  val transformDecodings = decodings.map {
+    case (decode, i) => (decode, decodeAs.indexOf(sft.getDescriptor(i).getLocalName))
+  }
+
+  override def read(kryo: Kryo, input: Input, typ: Class[SimpleFeature]): SimpleFeature = {
+    input.readInt(true) // discard version info, currently only one version
+    val id = input.readString()
+    val values = Array.ofDim[AnyRef](decodeAs.getAttributeCount)
+
+    transformDecodings.foreach { case (decode, i) =>
+      if (i == -1) decode(input) else values(i) = decode(input)
+    }
+    new KryoSimpleFeature(id, decodeAs, values)
+  }
+}
+
+object SimpleFeatureSerializer {
+
+  import scala.collection.mutable.Map
+
+  val VERSION = 0
+
+  val NULL_BYTE     = 0.asInstanceOf[Byte]
+  val NON_NULL_BYTE = 1.asInstanceOf[Byte]
+
+  type Encoding = (Output, SimpleFeature) => Unit
+  type Decoding = ((Input) => AnyRef, Int)
+
+  // encodings are cached per-thread to avoid synchronization issues
+  // we use soft references to allow garbage collection as needed
+  private val encodingsCache = new ThreadLocal[Map[String, SoftReference[Seq[Encoding]]]] {
+    override def initialValue = Map.empty[String, SoftReference[Seq[Encoding]]]
+  }
+  private val decodingsCache = new ThreadLocal[Map[String, SoftReference[Seq[Decoding]]]] {
```

Comment:
Ok, ok, if we are doing it *three or more places*....  I think we should extract the class I suggested...